DOC fix autocomplete documentation

tomMoral opened the pull request:
Does it fix your issue @mathurinm ?

mathurinm commented:
Partially ! Now some code is running ! But it raises an error:

```python
(base) ➜  lasso git:(main) ✗ benchopt run . -e -s s  (here I press tab and:)
Traceback (most recent call last):
  File "/home/mathurin/miniconda3/bin/benchopt", line 33, in <module>
    sys.exit(load_entry_point('benchopt', 'console_scripts', 'benchopt')())
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 1048, in main
    self._main_shell_completion(extra, prog_name, complete_var)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 1123, in _main_shell_completion
    rv = shell_complete(self, ctx_args, prog_name, complete_var, instruction)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 49, in shell_complete
    echo(comp.complete())
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 291, in complete
    completions = self.get_completions(args, incomplete)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 273, in get_completions
    return obj.shell_complete(ctx, incomplete)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 2384, in shell_complete
    results = self._custom_shell_complete(ctx, self, incomplete)
  File "/home/mathurin/workspace/benchopt/benchopt/cli/completion.py", line 58, in get_solvers
    benchmark = find_benchmark_in_args(args)
  File "/home/mathurin/workspace/benchopt/benchopt/cli/completion.py", line 48, in find_benchmark_in_args
    for b in args:
TypeError: 'Option' object is not iterable
```

tomMoral commented:
Which version of `click` do you have?

tomMoral commented:
Ok, the issue was with `click8.0`. Basically, pallets/click#1622 completely redesign the shell completion API.
From the release note:

> Redesign the shell completion system. :issue:`1484`, :pr:`1622`
>    -  Support Bash >= 4.4, Zsh, and Fish, with the ability for extensions to add support for other shells.
>    - Allow commands, groups, parameters, and types to override their completions suggestions.
>    - Groups complete the names commands were registered with, which can differ from the name they were created with.
>    - The autocompletion parameter for options and arguments is renamed to shell_complete. The function must take ctx, param, incomplete, must do matching rather than return all values, and must return a list of strings or a list of CompletionItem. The old name and behavior is deprecated and will be removed in 8.1.
>    - The env var values used to start completion have changed order. The shell now comes first, such as {shell}_source rather than source_{shell}, and is always required.
```

Maybe we should add some tests to make sure we detect this next time. A source of inspiration: https://github.com/pallets/click/blob/main/tests/test_shell_completion.py

mathurinm commented:
It works for benchopt run . -s and -d now ! :tada: 

I guess we need to do the same for `(base) ➜  lasso git:(main) ✗ benchopt plot . -f Traceback` and publish ? 

I realised we have `autocompletion.get_datasets()` and `BENCHMARK.get_datasets()`, those have nothing in common and the autocomplete one could be renamed or I missed something ?

tomMoral commented:
Good catch! I fixed this for all the arguments this time I think.

I also renamed `get_*` in `complete_*` to avoid the collapse with benchmark methods.

mathurinm commented on benchopt/cli/completion.py:
```diff
@@ -52,38 +52,38 @@ def find_benchmark_in_args(args):
 
-def get_output_files(ctx, args, incomplete):
+def complete_output_files(ctx, param, incomplete):
     "Auto-completion for datasets."
```
```suggestion
    "Auto-completion for output files."
```

mathurinm commented:
Sorry I could not look into it more, but `benchopt publish . -f` and `benchopt publish -f ` have problems in autocompletion:

```
(base) ➜  lasso git:(main) ✗ benchopt publish . -f Traceback (most recent call last):
  File "/home/mathurin/miniconda3/bin/benchopt", line 33, in <module>
    sys.exit(load_entry_point('benchopt', 'console_scripts', 'benchopt')())
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 1048, in main
    self._main_shell_completion(extra, prog_name, complete_var)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 1123, in _main_shell_completion
    rv = shell_complete(self, ctx_args, prog_name, complete_var, instruction)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 49, in shell_complete
    echo(comp.complete())
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 292, in complete
    out = [self.format_completion(item) for item in completions]
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 292, in <listcomp>
    out = [self.format_completion(item) for item in completions]
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 364, in format_completion
    return f"{item.type}\n{item.value}\n{item.help if item.help else '_'}"
AttributeError: 'PosixPath' object has no attribute 'type'

Error: Option '-f' requires an argument.
(base) ➜  lasso git:(main) ✗ benchopt publish -f Traceback (most recent call last):
  File "/home/mathurin/miniconda3/bin/benchopt", line 33, in <module>
    sys.exit(load_entry_point('benchopt', 'console_scripts', 'benchopt')())
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 1048, in main
    self._main_shell_completion(extra, prog_name, complete_var)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/core.py", line 1123, in _main_shell_completion
    rv = shell_complete(self, ctx_args, prog_name, complete_var, instruction)
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 49, in shell_complete
    echo(comp.complete())
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 292, in complete
    out = [self.format_completion(item) for item in completions]
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 292, in <listcomp>
    out = [self.format_completion(item) for item in completions]
  File "/home/mathurin/miniconda3/lib/python3.8/site-packages/click/shell_completion.py", line 364, in format_completion
    return f"{item.type}\n{item.value}\n{item.help if item.help else '_'}"
AttributeError: 'tuple' object has no attribute 'type'
```

mathurinm commented:
benchopt run . -s and -d still work

and now publish . -f works too !

```
(base) ➜  lasso git:(main) ✗ benchopt publish . -f outputs/benchopt_run_2021-12-07_20h35m11.csv
outputs/benchopt_run_2021-12-07_20h35m11.csv
outputs/benchopt_run_2021-11-17_20h03m14.csv
outputs/benchopt_run_2021-11-17_20h03m24.csv
outputs/benchopt_run_2021-11-05_08h30m38.csv
```
thanks for the fix @tomMoral
